enhancements/cassandra-configuration.md

```diff
@@ -0,0 +1,265 @@
+      
+      roles_validity_in_ms: 2000
+      permissions_validity_in_ms: 2000
+      credentials_validity_in_ms: 2000
+      
+      # How long before a node logs slow queries. Select queries that 
+      # take longer than this timeout to execute, will generate an
+      # aggregated log message, so that slow queries can be identified. 
+      # Set this value to zero to disable slow query logging. 
+      slow_query_log_timeout_in_ms: 500
+
+    jvm-options:
+      # Set the database to use 14 GB of Java heap
+      initial_heap_size: "14G"
+      max_heap_size: "14G"
+
+      additional-jvm-opts:
+        # As the database comes up for the first time, set system keyspaces to RF=3
+        - "-Dcassandra.system_distributed_replication_dc_names=dc1"
+        - "-Dcassandra.system_distributed_replication_per_dc=3"
+```
+
+I tried to include a number of properties for illustration. There are some however that 
+deserve further discussion.
+
+First, I want to point out that there are number of properties that we do not want to expose 
+including the following:
+
+* storage_port
+* ssl_storage_port
+* native_transport_port
+* seed_provider
+* listen_address
+* native_transport_port
+* rpc_address
+* data_files_directories
+* commitlog_directories
+
+There are a number of properties around authentication/authorization. we may not want to 
+directly expose properties like `authenticator` and `authorizer`. The auth subsystem is 
+pluggable which is great. I can use a custom `IAuthenticator` by specifying the class name 
+for `authenticator` and including it in Cassandra's classpath. Doing the latter is a 
+non-trivial exercise in Kubernetes.
+
+My example also included some auth cache settings:
+
+* roles_validity_in_ms
+* permissions_validity_in_ms
+* credentials_validity_in_ms
+
+These are not very good defaults. They can easily contribute to latency issues when auth is 
+enabled. We probably want to expose these but we also should override the defaults if auth 
+is enabled.
+
+Now let's talk about the slow query log which is enabled via the 
+`slow_query_log_timeout_in_ms` property. The slow query log can be really helpful for 
+debugging latency and performance issues. If this is enabled, we might want to consider 
+adding a logging sidecar for the debug log to make those logs more easily accessible. 
+
+Similarly, if the user wants to enable CDC we should make it easy to configure another 
+volume for it.
+
+For `jvm.options`, the big thing that immediately comes to mind is GC settings. 
+Depending on what collectors you use, e.g., ParNew/CMS, G1, etc., there are a number of 
+properties that you may want to configure. It should be enough for the user to simply 
+specify the collector to use and we should then provide sensible defaults. The user should 
+still have the ability to set specific GC properties.
+
+# K8ssandra
+## Topology
+```yaml
+k8ssandra:
+  clusterName: multi-rack-multi-dc
+  datacenters:
+  - name: dc1
+    racks:
```

Comment:
What about the number of nodes in each datacenter? How will `k8ssandra` know how many nodes to assign to each datacenter?
In `CassandraDatacenter.yaml` we have the `size` parameter which translates to the number of Cassandra nodes in the datacenter. AFAIK the cass-operator will distribute the C* node evenly between the racks.
Is the plan of `k8ssandra` to do the same with distributing C* nodes between DCs?

Reply:
Also, please consider the following use case in which I am working on (comments appreciated). In this case it would be nice if we could specify the number of C* nodes per rack.

# k8ssandra with disaggregated storage

## Requirements

- The number of virtual Cassandra racks (failure domains) must be **less than or equal** to the number of disaggregated storage failure domains
- The number of k8s compute physical failure domains can be much higher

## Working Example

**Compute (k8s)**
- 10 racks (k8s failure domains)
- 30 nodes per rack - total 300 nodes

**Disaggregated Storage**
- 3 failure domains

**Storage (Cassandra)**
- 3 virtual racks
- 100 nodes per rack
  - 98 active C* nodes per rack
  - 2 backup C* nodes per rack

### k8s cluster configuration

#### failure domains mapping

Each physical node belongs to a physical failure domain.
In order to answer the first requirement, we map each physical node to a virtual Cassandra rack by grouping physical racks to virtual racks and assigning each k8s node the label `failure-domain.beta.kubernetes.io/zone` accordingly:

| physical rack | physical nodes | virtual rack  | failure-domains.beta.kubernetes.io/zone |
|---|---|---|---|
| rack1 | 1-30 | rack1 | "rack1" |
| rack2 | 31-60 | rack1 | "rack1" |
| rack3 | 61-90 | rack1 | "rack1" |
| rack4 | 91-120 | rack2 | "rack2" |
| rack5 | 121-150 | rack2 | "rack2" |
| rack6 | 151-180 | rack2 | "rack2" |
| rack7 | 181-210 | rack3 | "rack3" |
| rack8 | 211-240 | rack3 | "rack3" |
| rack9 | 241-270 | rack3 | "rack3" |
| rack10 | 271-280 | rack1 | "rack1" |
| rack10 | 281-290 | rack2 | "rack2" |
| rack10 | 291-300 | rack3 | "rack3" |

### k8ssandra cluster configuration

k8ssandra configuration will include the virtual racks we defined earlier (according to the [k8ssandra topology configuration design doc](https://github.com/k8ssandra/k8ssandra/blob/c5410f47413b69e2b57777b75cf68f9d6fe15a7a/enhancements/cassandra-configuration.md#topology-1))

```yaml
k8ssandra:
  clusterName: disaggregated-storage
  size: 294
  datacenters:
  - name: dc1
    racks:
    - name: rack1
      zone: rack1
    - name: rack2
      zone: rack2
    - name: rack3
      zone: rack3
```

This will translate to the following cass-operator CassandraDatacenter.yaml:

```yaml
apiVersion: cassandra.datastax.com/v1beta1
kind: CassandraDatacenter
metadata:
  name: dc1
spec:
  clusterName: disaggregated-storage
  serverType: cassandra
  serverVersion: 3.11.7
  managementApiAuth:
    insecure: {}
  size: 294
  racks:
  - name: rack1
    zone: rack1
  - name: rack2
    zone: rack2
  - name: rack3
    zone: rack3
  ...
```

Which will eventually translate to 3 stateful sets, one per Cassandra virtual failure domain - rack1,rack2 and rack3. The following is the rack1 statefulset example:

disaggregated-storage-dc1-sts-rack1.yaml
```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: cassandra
spec:
  serviceName: "disaggregated-storage-dc1-service"
  replicas: 98
  template:
    metadata:
      labels:
        app: cassandra
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
              - matchExpressions:
                - key: failure-domain.beta.kubernetes.io/zone
                  operator: In
                  valures:
                  - rack1
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
              matchExpressions:
                - key: app
                  operator: In
                  values:
                    - cassandra
              topologyKey: failure-domains.beta.kubernetes.io/zone
```

### k8s scheduling

The above StatefulSet NodeAffinity manifest tells Kubernetes to schedule each replica to one of the k8s nodes with label `failure-domain.beta.kubernetes.io/zone=rack1`. The podAntiAffinity rule tells Kubernetes not to place more than one instance of the application in the same k8s node.
The result - each C* node is running on a dedicated k8s node with the rack1 label.

Due to the 10 physical racks to 3 virtual racks mapping, Kubernetes is free to schedule each of the 98 C* nodes of each rack to any one of 100 k8s nodes with that rack's label.

Since each node in virtual rack1 belongs to physical rack1, rack2, or rack3 - it is also guaranteed that each C* node will run on one of the physical failure domains.

### k8s volume provisioning

Each C* node (pod) in the statefulset defines a volume which is used for the C* storage using the predefined `server-storage` storage class.
Kubernetes will provision the volume with the topology information, and the CSI plugin will provision the volume in the corresponding disaggregated storage failure doma, according to the rack number.


### Failure and recovery

Recall that we pre-allocated 100 k8s nodes per Cassandra rack, but set the number of nodes in each rack to 98.
This leaves us with 2 backup nodes for each rack which are used for failure recovery.

#### C* node (pod) failure

Since every C* node runs on a dedicated k8s node, when a single pod fails it will be restarted on the same node.

#### k8s node failure

On node failure, k8s will restart the failed C* node (pod) on one of the 2 backup nodes of that virtual rack.
The PVC of the failed node will be attached to the backup node.

#### Physical compute rack failure

We can't cover up for a whole rack failure in this Cassandra deployment example, so this is left up for Cassandra to handle.
In case of RF=3 murmor3 configuration,  Cassandra will still have 2 replicas in rack2 and rack3 so it will still have quorum.

**Edge case - physical rack10 failure**

Physical rack10 contains 10 nodes from Cassandra failure domain. This means that 10 nodes in each virtual Cassandra rack are down, so in case that Cassandra distributed the 3 replicas between those specific nodes there will be no availability.

To prevent this edge case, instead of splitting all nodes from rack10 between the virtual racks - we can assign them to one of the racks.
This, However, is not possible currently with cass-operator / k8ssandra.

Reply:
**Finer-graned scheduling support to k8ssandra and cass-operator**
Another way to solve this is with the `allowMultipleNodesPerWorker` configuration option from cass-operator - instead of using "backup k8s nodes", we can allow 2 C* nodes to run on the same k8s worker node in some occasions.
Currently, `allowMultipleNodesPerWorker` simply removes the podAntiAffinity rule so that Kubernetes can theoretically schedule much more than 2 C* nodes in the same k8s node, but we can use the new [pod-topology-spread-constraints](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/) to fine-tune the behavior.

Reply:
> What about the number of nodes in each datacenter? How will `k8ssandra` know how many nodes to assign to each datacenter?

We use a default value of 1. The user will have to set it more generally though.

> In `CassandraDatacenter.yaml` we have the `size` parameter which translates to the number of Cassandra nodes in the datacenter. AFAIK the cass-operator will distribute the C* node evenly between the racks.
> Is the plan of `k8ssandra` to do the same with distributing C* nodes between DCs?

cass-operator does a really good job of trying to ensure there are balanced racks, but it cannot be guaranteed based on what the user specifies. Rather than using the `size` property, I prefer to use something like `nodesPerRack` and `numRacks` which will allow us to ensure balanced racks.

Reply:
Actually you don't need `numRacks` if the user anyway has to define all racks names and labels (and he should). The `nodesPerRack` is also not really needed now that I think about it again - I originally suggested this cause I thought that in some occasions users may prefer to have more nodes in some racks, but after reading more about the caveats of unbalanced racks I take it back. So you can safely ignore my long comment (I can delete it or you can `resolve` it).